### PR TITLE
Refactor of LocalizedNumber

### PR DESCRIPTION
**Pull request by oeoeaio**
#### What? Why?

I did a bit of a refactor of the LocalizedNumber module created by @ltrls to allow more flexibility in the use of decimal separators for numbers entered by users (#1831).

There are a few outstanding issues from #1831 that still need to be addressed, and so they can be added here as well.

#### What should we test?

Functionality should not change, this is just a bit of a tidy up of the code.

#### Release notes

No release notes necessary.

### CODE REVIEW

File: lib/spree/localized_number.rb [hunk 1]
```diff
@@ -37,6 +29,31 @@ def localize_number(*attributes)
...
+          self[attribute] = validated_number
+        else
+          old_setter.bind(self).call(validated_number)
+        end
+      end
+    end
+
+    def self.validated_number(number)
+      return parse(number) if valid_localizable_number?(number)
+      nil
```
Comment (sauloperez) [inline lib/spree/localized_number.rb]: I believe this nil would be implicitly returned by Ruby

File: lib/spree/localized_number.rb [hunk 2]
```diff
@@ -37,6 +29,31 @@ def localize_number(*attributes)
       end
     end
 
+    def override_setter_for(attribute)
+      setter = "#{attribute}="
+      old_setter = instance_method(setter) if table_exists? && !column_names.include?(attribute.to_s)
```
Comment (sauloperez) [inline lib/spree/localized_number.rb]: I fail to see what's this case. An AR attribute not baked by a database table column?

File: lib/spree/localized_number.rb (hunk 1 above)
Comment (oeoeaio) [reply lib/spree/localized_number.rb]: 👍  that's true

File: lib/spree/localized_number.rb (hunk 2 above)
Comment (oeoeaio) [reply lib/spree/localized_number.rb]: I think it is for preferences? Spree's calculators all share the same table, but they have preferences which are stored in a separate table, using the old setter means that we still set the preference properly. @ltrls should be able to confirm...

**Comment by daniellemoorhead** [comment]
@oeoeaio is this something that is still on your radar? Have moved it to In Dev as it's not ready for review again, please close if it's no longer relevant.

#githubcull2018

**Comment by daniellemoorhead** [comment]
Hey @oeoeaio what's the plan of attack with this PR? Can it be handed over to someone else to complete, or closed off as no longer needed given #853 is now closed?

**Comment by myriamboure** [comment]
Note: the two remaining issues that we opened after we merge the fix for decimals are:
- https://github.com/openfoodfoundation/openfoodnetwork/issues/1970
- https://github.com/openfoodfoundation/openfoodnetwork/issues/1971
It's more UX issues but introduced by this decimal work so I guess that's what you talk about @oeoeaio 
What do we want to do with that ? Shall we open an icebox and put that there, as now users are not so much annoyed by errors when they put , instead of . ;-) So I think it can wait. Opinion welcome @sauloperez

**Comment by daniellemoorhead** [comment]
This isn't being worked on at all right now so I'm pushing this back to the backlog. @oeoeaio feel free to pick it up once you're free from subs stuff.

**Comment by daniellemoorhead** [comment]
@oeoeaio I've unassigned you from this PR as I assume you won't be doing it anytime soon and someone else can pick it up to finish it.

Feel free to reassign yourself if you want to finish it off.

**Comment by daniellemoorhead** [comment]
Actually, I'm going to close it for now.

@mkllnk @Matt-Yorkley @luisramos0 @sauloperez if you believe this should be completed and merged then feel free to reopen.

**Pull request closed without merging**